framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,538 @@
+				if (m->fd < 0) {
+					return NULL;
+				}
+				m->state = MEDIA_STATE_RECORDING;
+				return m;
+			} else if (play_type == MEDIA_PLAY_TYPE_NETWORK) {
+				// TODO: Network streaming support
+			}
+			return NULL;
+		}
+	}
+	return NULL;
+}
+
+void media_close(media_t *m)
+{
+	if (m == NULL) {
+		return;
+	}
+	if (m->fd >= 0) {
+		close(m->fd);
+	}
+	free(m);
+	return;
+}
+
+
+int on_media_state_playing(struct pcm *pcmout, media_t *m, char *buffer, unsigned int buffer_size)
+{
+	int readed;
+	int remain;
+	int ret;
+	readed = read(m->fd, buffer, buffer_size);
+	if (readed) {
+		remain = pcm_bytes_to_frames(pcmout, readed);
+		while (remain > 0) {
+			ret = pcm_writei(pcmout, buffer + readed - pcm_frames_to_bytes(pcmout, remain), remain);
+			if (ret <= 0) {
+				return MEDIA_ERROR_PCM_WRITE;
+			} else {
+				remain -= ret;
+			}
+		}
+	}
+
+	return readed;
+}
+
+int on_media_state_recording(struct pcm *pcmin, media_t *m, char *buffer, unsigned int buffer_size)
+{
+	int readed;
+	int remain;
+	int ret;
+
+	readed = pcm_readi(pcmin, buffer, pcm_bytes_to_frames(pcmin, buffer_size));
+	if (readed) {
+		remain = pcm_frames_to_bytes(pcmin, readed);
+		while (remain > 0) {
+			ret = write(m->fd, buffer + pcm_frames_to_bytes(pcmin, readed) - remain, remain);
+			if (ret < 0) {
+				return MEDIA_ERROR_WRITE_TO_FILE;
+			} else {
+				remain -= ret;
+			}
+		}
+	}
+	return readed;
+}
+
+/***
+ * Thread routine
+ */
+int threadfunc_audio_playing(void *args)
+{
+	char *buffer;
+	unsigned int buffer_size;
+	struct pcm *pcmout;
+	struct pcm_config config;
+	dq_entry_t *node;
+	int ret;
+#ifndef CONFIG_AUDIO_MULTI_CARD
+	struct pcm *pcmin;
+	bool playing = false;
+	bool recording = false;
+	buffer = NULL;
+	buffer_size = 0;
+	pcmout = NULL;
+	pcmin = NULL;
+#endif
+	// Configuration
+	config.channels = 2;
+	config.rate = 16000;
+	config.period_size = CONFIG_AUDIO_BUFFER_NUMBYTES;
+	config.period_count = CONFIG_AUDIO_NUM_BUFFERS;
+	config.format = PCM_FORMAT_S16_LE;
+	config.start_threshold = 0;
+	config.stop_threshold = 0;
+	config.silence_threshold = 0;
+
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	pcmout = pcm_open(0, 0, PCM_OUT, &config);
+	buffer_size = pcm_get_buffer_size(pcmout);
+	buffer = (char *)malloc(buffer_size);
+#endif
+
+#ifdef CONFIG_AUDIO_MULTI_CARD
+	printf("audio_playing thread started. Buffer size: %d\n", buffer_size);
+#else
+	printf("audio thread started.\n");
+#endif
+
+	g_playing_live = true;
+
+	sem_wait(&g_sem_playing);
+	while (g_playing_live) {
+		pthread_mutex_lock(&g_mutex_playing_q);
+		for (node = (dq_entry_t *)dq_peek(&g_playing_q); node && g_playing_live;) {
+			switch (((media_t *)node)->state) {
+			case MEDIA_STATE_PLAYING:
+#ifndef CONFIG_AUDIO_MULTI_CARD
+				if (recording) {
+					node = dq_next(node);
+					continue;
+				}
+				if (!playing) {
+					pcmout = pcm_open(0, 0, PCM_OUT, &config);
+					buffer_size = pcm_get_buffer_size(pcmout);
+					buffer = (char *)malloc(buffer_size);
```

Comment:
When config_audio_multi_card is enabled, isn't it dupplicated?

Reply:
No, it is in #ifndef macro. so the code block is active only when config_audio_multi_card is not defined